plugin/plugin.go

```diff
@@ -363,6 +373,231 @@ func RegisterShutdown(name string, s Shutter) error {
+// pointers to its child items. Getting the child pointers requires a little
+// more arithmetic than in C as we can't just increment a pointer to get the
+// next array item
+func getOconfigChildren(oconfig *C.oconfig_item_t) (output []*C.oconfig_item_t) {
+	start := unsafe.Pointer(oconfig.children)
+	size := unsafe.Sizeof(*oconfig.children)
+	for i := 0; i < int(oconfig.children_num); i++ {
+		child := (*C.oconfig_item_t)(unsafe.Pointer(uintptr(start) + size*uintptr(i)))
+		output = append(output, child)
+	}
+	return
+}
+
+// checkMatchable returns an error if the provided source cannot be copied to
+// the target, or nil if it's OK
+func checkMatchable(oconfig *C.oconfig_item_t, target reflect.Value) error {
+	if oconfig.children_num > 0 { // if it has children we need a corresponding struct
+		if target.Kind() != reflect.Struct {
+			return fmt.Errorf("did not have corresponding struct")
+		}
+	}
+	if oconfig.values_num > 1 { // if it has multiple values we need a slice
+		if target.Kind() != reflect.Slice {
+			return fmt.Errorf("did not have corresponding slice")
+		}
+	}
+	if oconfig.values_num == 1 { // if it has a single value we need either a string, int or bool
+		valType, err := C.go_get_value_type(oconfig, C.int(0))
+		if err != nil {
+			return fmt.Errorf("unable to determine type of value")
+		}
+		switch valType {
+		case C.int(0): //string
+			if target.Kind() != reflect.String {
+				return fmt.Errorf("found string, plugin expects %v", target.Kind().String())
+			}
+		case C.int(1): // number/double/float64
+			if target.Kind() != reflect.Float64 {
+				return fmt.Errorf("found number, plugin expects %v", target.Kind().String())
+			}
+		case C.int(2): // boolean
+			if target.Kind() != reflect.Bool {
+				return fmt.Errorf("found boolean, plugin expects %v", target.Kind().String())
+			}
+		default:
+			return fmt.Errorf("unsupported Collectd config item type")
+		}
+	}
+	return nil
+}
+
+// checkUsable iterates over a provided reflect.Value and attempts to make sure
+// it A) contains only string, bool, or float64 fields, or slices or structs of
+// these, and B) contains only settable, exported fields
+// Unfortunately as we start with a reflect.Value and no reflect.Type we can't
+// actually report the name of the failing field. Fortunately this should be a
+// plugin error only (not possible for a user to trigger) and resolved during
+// notmal testing.
+func checkUsable(target reflect.Value) error {
+	switch target.Kind().String() {
+	case "struct":
+		for i := 0; i < target.NumField(); i++ {
+			err := checkUsable(target.Field(i))
+			if err != nil {
+				return fmt.Errorf("in struct: %v", err)
+			}
+		}
+	case "slice":
+		if target.CanSet() != true {
+			return fmt.Errorf("unsettable value (unexported field?)")
+		}
+		// Get an Interface of the slice, extract a field and get the field's type
+		switch reflect.TypeOf(target.Interface()).Elem().Kind() {
```

Comment:
Looks like I tied myself in knots trying to think that one through. `target.Type()` will return a dynamic reflect.Type instead of a reflect.Kind, but 
```go
target.Type().Elem().Kind()
``` 
(get Type, then get it's element Type, then get the Kind of that Type) is much simpler. Fixed in next commit.